pkg/networkpolicy/controller.go

```diff
@@ -258,28 +259,47 @@ func (c *Controller) Run(ctx context.Context, workers int) error {
 	return nil
 }
 
-// syncIptablesRules adds the necessary rules to process the first connection packets in userspace
+// syncNFTablesRules adds the necessary rules to process the first connection packets in userspace
 // and check if network policies must apply.
-// --queue-bypass is on other NFQUEUE option by Florian Westphal.
-// It change the behavior of a iptables rules when no userspace software is connected to the queue.
-// Instead of dropping packets, the packet are authorized if no software is listening to the queue.
-func (c *Controller) syncIptablesRules() {
-	if err := c.ipt.InsertUnique("filter", "FORWARD", 1, "-m", "conntrack", "--ctstate", "NEW", "-j", "NFQUEUE", "--queue-bypass", "--queue-num", strconv.Itoa(c.nfqueueID)); err != nil {
-		klog.Infof("error syncing iptables rule %v", err)
-	}
+func (c *Controller) syncNFTablesRules(ctx context.Context) {
+	rule := fmt.Sprintf("ct state new  queue to %d", c.nfqueueID)
+
+	tx := c.nft.NewTransaction()
+	tx.Add(&knftables.Table{
+		Comment: knftables.PtrTo("rules for kubernetes NetworkPolicy"),
+	})
 
-	if err := c.ipt.InsertUnique("filter", "OUTPUT", 1, "-m", "conntrack", "--ctstate", "NEW", "-j", "NFQUEUE", "--queue-bypass", "--queue-num", strconv.Itoa(c.nfqueueID)); err != nil {
-		klog.Infof("error syncing iptables rule %v", err)
+	for _, hook := range []knftables.BaseChainHook{knftables.InputHook, knftables.ForwardHook, knftables.OutputHook} {
+		chainName := string(hook)
+		tx.Add(&knftables.Chain{
+			Name:     chainName,
+			Type:     knftables.PtrTo(knftables.FilterType),
+			Hook:     knftables.PtrTo(knftables.ForwardHook),
+			Priority: knftables.PtrTo(knftables.FilterPriority),
+		})
+		tx.Flush(&knftables.Chain{
+			Name: chainName,
+		})
+		tx.Add(&knftables.Rule{
+			Chain: chainName,
+			Rule:  rule,
+		})
+	}
+
+	if err := c.nft.Run(ctx, tx); err != nil {
+		klog.Infof("error syncing nftables rules %v", err)
 	}
 }
 
-func (c *Controller) cleanIptablesRules() {
-	if err := c.ipt.Delete("filter", "FORWARD", "-m", "conntrack", "--ctstate", "NEW", "-j", "NFQUEUE", "--queue-bypass", "--queue-num", strconv.Itoa(c.nfqueueID)); err != nil {
-		klog.Infof("error deleting iptables rule %v", err)
-	}
+func (c *Controller) cleanNFTablesRules(ctx context.Context) {
+	tx := c.nft.NewTransaction()
+	// Add+Delete is idempotent and won't return an error if the table doesn't already
+	// exist.
+	tx.Add(&knftables.Table{})
+	tx.Delete(&knftables.Table{})
```

Comment:
shouldn't we have to delete `chainName` ?

Reply:
No, we delete the entire table. That's how you do things in nftables-land. You put your own rules in your own table, so then you're done with it, you can just delete the whole table rather than needing to think about every specific thing you created.

(The knftables API is a little bit non-obvious here: the table name and family are set at `nft` creation time (since otherwise you'd have to specify them on every operation), so there's only one nftables `table` that you can be referring to here and so you don't need to name it in the `knftables.Table` object.)